core/rfb.js

```diff
@@ -482,60 +456,66 @@
                     }
                     break;
 
                 case 'connecting':
+                    // A previous connect may asynchronously cause a connection
+                    // so make sure we are still closed.
+                    this._cleanupSocket();
                     this._init_vars();
+
+                    // WebSocket.onopen transitions to the RFB init states
                     this._connect();
-                    // WebSocket.onopen transitions to 'ProtocolVersion'
                     break;
 
                 case 'disconnecting':
+                    // WebSocket.onclose transitions to 'disconnected'
+                    this._cleanupSocket();
+
                     this._disconnTimer = setTimeout(function () {
                         this._fail("Disconnect timeout");
                     }.bind(this), this._disconnectTimeout * 1000);
 
                     this._print_stats();
-
-                    // WebSocket.onclose transitions to 'disconnected'
-                    break;
-
-                case 'failed':
-                    if (oldstate === 'disconnected') {
-                        Util.Error("Invalid transition from 'disconnected' to 'failed'");
-                    } else if (oldstate === 'connected') {
-                        Util.Error("Error while connected.");
-                    } else if (oldstate === 'init') {
-                        Util.Error("Error while initializing.");
-                    }
                     break;
 
                 default:
                     // No state change action to take
             }
-
-            if (oldstate === 'failed' && state === 'disconnected') {
-                // do disconnect action, but stay in failed state and
-                // keep the previous status message
-                this._rfb_connection_state = 'failed';
-                this._onUpdateState(this, state, oldstate);
-            } else {
-                this._onUpdateState(this, state, oldstate, statusMsg);
-            }
-
-            // Make sure we transition to disconnected
-            if (state === 'failed') {
-                this._updateConnectionState('disconnected');
-            }
         },
 
         _fail: function (msg) {
-            this._updateConnectionState('failed', msg);
+            switch (this._rfb_connection_state) {
+                case 'disconnecting':
+                    Util.Error("Error while disconnecting: " + msg);
+                    break;
+                case 'connected':
+                    Util.Error("Error while connected: " + msg);
+                    break;
+                case 'connecting':
+                    Util.Error("Error while connecting: " + msg);
+                    break;
+                default:
+                    Util.Error("RFB error: " + msg);
+                    break;
+            }
+            this._updateConnectionState('disconnected', msg);
```

Comment:
I like this, but I think this should actually probably be "disconnecting" (because we haven't yet closed the WebSocket).  I think the old code (which jumped from "failed" to "disconnected") was either unclear or "incorrect" (in how it labeled states).